# Classifying Organisations
## Business and Non-Business Classification

One of the interesting aspects of the Gateway to Research data is that it includes a significant number of non-academic organisations. We suspect that these are, in most cases, likely to be private companies that have received funding from Innoate UK. It may also include organisations that have collaborated with universities on specific projects.

Our first appraoch to identifying academic organisations and non-academic organisations will be to use keywords. For instance, companies can be identified via words such as 'ltd', 'plc' or 'inc', whilst academic organisations will often contain 'university' or 'school'. We also include 'council' in our first round of words identifying non-business organisations. 

In [3]:
import chwrapper
import json
import numpy as np
import os
import pandas as pd
import psycopg2
import ratelim
import requests

from sqlalchemy import create_engine

In [256]:
pd.options.display.max_rows = 250

In [10]:
def check_type(c_str, company=True, custom_list=None):
    """Takes a string and checks whether it is a name representing a company or not.
    Set company=False to check whether the string DOES NOT represent a company."""
    words = set(c_str.lower().split())
    return is_company(words) if company else is_other(words, custom_list)
 
def is_company(words):
    """Takes words and compares them to a list of company identifiers. Returns True if
    any of the words match those in the identifiers list."""
    identifiers = set(['ltd.',
                       'ltd',
                       'limited',
                       'plc',
                       'plc.',
                       'inc',
                       'inc.',
                       'company',
                       'corporation'])
    return bool(words.intersection(identifiers))

def is_other(words, custom_list=None):
    """Takes words and compares them to a list of otber identifiers (denoting
    academic or public organisation keywords. Returns True if any of the words
    match those in the identifiers list."""
    if not custom_list:
        identifiers = set(['university',
                           'uni',
                           'universities',
                           'school',
                           'sch',
                           'faculty',
                           'council',
                           'nhs',
                           'academic',
                           'science',
                           'sciences',
                           'sci',
                           'office',
                           'institute'])
    else:
        identifiers=set(custom_list)
        
    return bool(words.intersection(identifiers))

def print_stats(df):
    # Number of organisations that are other (academic, public sector etc.)
    print('Other:\t\t{}'.format(len(df[df.other == True])))

    # Number of organisations that aren't companies
    print('Companies:\t{}'.format(len(df[df.company == True])))

    # Unclassified
    print('Unclassified:\t{}'.format(len(df[(df.company == False) &
                                        (df.other == False)])))

    # Clashes
    print('Clashes:\t{}'.format(len(df[(df.company == True) &
                                       (df.other == True)])))
    
def unpack_address(address_series):
    """Unpacks GtR nested addresses. Returns a pandas.Series object"""
    s = address_series.apply(lambda x: x.get('address', None)).apply(unpack_list)
    s = s.apply(dict_2_string)
    return s

def check_postcodes(row):
    """Check a dataframe rows postcode entry against that in a Companies House JSON return"""
    return ch_postcodes(row.ch_json)

def gtr_postcodes(address_series):
    """Returns a postcode entry from a row in a GtR derived address series"""
    s = address_series.apply(lambda x: x.get('address', None)).apply(unpack_list)
    return s.apply(val_from_dict, key='postCode')

def ch_postcodes(ch_json):
    """Returns all postcode entries from a Companies House JSON return"""
    items = ch_json.get('items', None)
    if items is None:
        return None
    else:
        return [x.get('address').get('postal_code') for x in items]
    
def oc_postcodes(oc_json):
    """Returns all postcodes from an Open Corporates company search results document as a list"""
    try:
        companies = oc_json.get('results').get('companies')
    except AttributeError:
        companies = False
    
    if companies:
        l = []
        for company in companies:
            try:
                l.append(company.get('company').get('registered_address').get('postal_code'))
            except AttributeError:
                l.append(None)
        return [x for x in l if x is not None]
    else:
        return []
    
def val_from_dict(d, key):
    try:
        return d.get(key, None)
    except AttributeError:
        return None

def unpack_list(l, ind=0):
    try:
        return l[ind]
    except IndexError:
        return None

def dict_2_string(d):
    try:
        return ' '.join("{!s}".format(val) for val in d.values())
    except AttributeError:
        return None

def org_name_counter(df):
    """Counts of each string occurence in name column"""
    return df.name.value_counts()

def unclassified(df):
    """Returns a df of other and company columns where both are False in df"""
    return df[(df.company == False) & (df.other == False)]

@ratelim.patient(600, 300)
def search_org(org_name, s=None):
    """Search companies house for an organisation by name (org_name)"""
    if not s:
        r = chwrapper.Search().search_companies(org_name)
    else:
        r = s.search_companies(org_name)
    return r.json()

@ratelim.greedy(1, 2)
def search_open_corporates(org_name, country_code=None, api_token=None):
    """Search the Open Corporates database using a string"""
    base_url = 'https://api.opencorporates.com/companies/search'
    params = {'q': org_name,
              'country_code': country_code,
              'api_token': api_token}
    
    r = requests.get(base_url, params=params)
    return r.json()

In [66]:
r = search_open_corporates("Nesta", country_code='gb', api_token=oc_key)

In [67]:
r['results']

{'companies': [{'company': {'branch_status': None,
    'company_number': '08506131',
    'company_type': 'Private Limited Company',
    'created_at': '2013-04-26T11:12:10+00:00',
    'current_status': 'Dissolved',
    'dissolution_date': '2014-09-30',
    'inactive': True,
    'incorporation_date': '2013-04-26',
    'jurisdiction_code': 'gb',
    'name': 'BWB (NO.3) LIMITED',
    'opencorporates_url': 'https://opencorporates.com/companies/gb/08506131',
    'previous_names': [{'company_name': 'NESTA LIMITED',
      'end_date': '2013-07-22',
      'start_date': '2013-04-26'}],
    'registered_address': {'country': 'United Kingdom',
     'locality': 'London',
     'postal_code': 'EC4M 6YH',
     'region': None,
     'street_address': '2-6 Cannon Street'},
    'registered_address_in_full': '2-6 Cannon Street, London, EC4M 6YH',
    'registry_url': 'https://beta.companieshouse.gov.uk/company/08506131',
    'restricted_for_marketing': None,
    'retrieved_at': '2016-03-28T07:04:06+00:00',
  

In [4]:
sql_str = """
    SELECT id, name, addresses 
      FROM gtr.organisations
"""

# Read in config file with DB params
with open('../scripts/config.json') as f:
    conf = json.load(f)
    
# Define a connection string
conn_string = 'host={} dbname={} user={} password={}'.format(conf.get('host'),
                                                             conf.get('database'),
                                                             conf.get('user'),
                                                             conf.get('passw'))

# Create a connection object
conn = psycopg2.connect(conn_string)

In [5]:
# Dataframe from SQL data
df = pd.read_sql(sql_str, conn)

# Lowercase all the string names for easier matching
df['name'] = df.name.map(lambda x: x.lower())

# Column of boolean values (true if is company)
df['company'] = df.name.apply(check_type)

# Column of boolean values if (true if is not company)
df['other'] = df.name.apply(check_type, company=False)

# Get postcodes
df['postcodes'] = gtr_postcodes(df.addresses)

# Unpack the addess values
df['addresses'] = unpack_address(df.addresses)

print_stats(df)

Other:		4214
Companies:	11357
Unclassified:	10063
Clashes:	69


So, we have 4,214 other organisations, 11,357 companies and 10,063 unclassified organisations. We'll take a closer look at the unclassified organisations first, to see whether we can add to our classification lists. We only have 69 clashes, which is great to see, as it shows there is a clear seperation of nomenclature betweent the types of organisations at this point.

In [6]:
# Counts of each string occurence in name column
unclassified(df)

id  \
4      B76FE143-1E07-40B0-8932-028207296A78   
9      BA59F886-138D-41EB-B9B1-0C79E45C61D3   
11     C059DB15-1836-430B-A993-0759206ABBB8   
16     026F1607-31FD-4EE2-B366-1470B441EAC7   
20     C8A76B9E-4F39-490B-BCCF-009F592A756C   
28     D42BD99F-AD31-43D6-8FBE-0B96F1E266D3   
30     CF4A4186-2B62-41F1-BC4E-0151FE4C5784   
33     D5F4323F-A8DF-4869-B0F0-0EB45A7A7F95   
36     D72C440C-6E4E-4FEE-94D9-0652BB970C3B   
39     D819E77B-1EC0-4952-BA54-04F155598E41   
41     D9A1CFB4-D277-415E-AEE8-04CDC3D9198B   
42     DA8D00D6-71CC-4A8C-98DD-07960A596938   
43     DAAA3102-13FC-426C-9744-05111A9E43C9   
47     D460C34A-963F-491D-84E9-C7A973167CAD   
48     DCDC3741-03C2-437B-9C9B-0C39BB333F79   
49     DDE9FF86-611C-4727-BA3C-0426362D27A6   
51     E600E1C3-69AB-444F-B2E8-0BA8DB05C777   
54     EEC4B6C9-3087-4A0F-8778-09A011F6E379   
55     F1BF3198-2CFE-4C1D-850B-01EE17A7BAA6   
58     ECE0A5CB-B026-4C0A-B3B5-E26868C9BC5F   
59     F514689F-FD91-417C-861B-04BA585C5745   
61     F67BC0E5-F86A-47F2-B413-04D87295AF36   
63     E7B24702-D2CD-47B1-8739-0A00564E54A3   
68     EB8C9378-D6BB-4B33-8C62-0422082C43DC   
72     ECE0B44A-F1ED-413E-A20B-072B71B90A89   
75     FACD958B-AABD-455E-B51B-06F1803C5AF5   
77     17E48B0C-106F-4A9C-BFB9-59DFEEF2FCE2   
79     FFA0AE43-4E62-4D7C-B74A-0E0E86CDC284   
80     004F3FEE-B008-4817-9F59-101379EB6E91   
86     FBC908A0-AC3F-4A64-9B2A-0289ACBB5B9B   
88     FC6F0FD2-411F-4B2F-BECB-0DC46606958F   
90     FDE5B22A-9282-462C-80BF-0526B141CB9C   
92     0CBC3ED3-19D7-4877-8CA2-1245FB1D11B7   
94     0D5B2A3B-6372-4D97-91E4-17C01E6CE6C2   
99     FB408437-BE40-4848-849E-CD852D9408C5   
101    20199213-451B-4427-ADCC-15156E3DFDD2   
103    20CF07A3-2453-4A36-86DC-16FB025FCAA6   
105    18D04611-ABAC-43F6-A9C8-170E52A24F1E   
106    1B794BDD-D765-43FE-A98B-188C0734CFBC   
107    1D1341FA-F1F1-4046-8D33-13B99C42039F   
108    1501C110-4B18-493D-B738-07A28F30D768   
109    16D0EA4F-1295-4BBA-BE0E-135EB07EBB03   
110    122AFEA5-60CB-4652-9D7A-1426D09888E3   
111    123ECC18-50AF-4836-A79D-10C5F0F82041   
112    12A768C9-1D27-4F24-9FB0-1B430FF3880C   
115    27170BAD-BC94-44F1-9E3D-152745F86A7F   
116    2C22AA90-B3EE-41D3-BC9F-16275AA9A006   
119    FCA00D51-294E-4B74-94C4-185F8F1668FC   
127    38379F1F-79F7-4103-8474-1576B8BD09E3   
128    3895593A-3052-40BF-BB19-0FB955E556B3   
129    38A541E3-D613-4A6E-AAFE-191C036D8A3F   
132    31AF7C42-D7C4-47FF-A5D8-19AADBAD30A9   
133    32F2829F-8D62-4F50-BE9C-11F9D2150CE8   
139    3F3A6416-58DA-4CA8-9E00-114281B708A4   
140    3FCA55FB-0C5D-4375-94D6-119F7C1494E7   
141    3FE02922-0FA0-4843-9659-117D903FC811   
142    40430561-0F09-4744-891D-0ED15C90258C   
149    57B5D407-77E5-487E-A78B-15D7DD22D66F   
159    6108311C-4917-40B0-8A5C-11C0143F791B   
163    4FBC169D-DA4F-45EC-AE73-0DCE64F85044   
164    50F395D3-A6C9-4E49-B17B-184591A6FCE3   
169    4D6A8C0E-1BC0-4091-80A6-1A5E0463DCB4   
171    4DE64C44-6643-4910-B9DC-122EAA5168E3   
172    450B17F6-6803-40A2-957A-0DDFA9B3C214   
173    454A7544-43CA-4EAC-B545-1599E0DD4581   
178    D7F4F462-0518-4784-908A-D12633C139B3   
179    0130A676-ADB1-416C-9F35-09D8416EDFFB   
181    04298A94-DEF7-4D8F-B575-05ADD0B9DE78   
182    044D767A-FF93-47E8-A691-0895AA38D3BF   
186    0EBD7BC9-DEBF-4E8F-A070-096D5DCB7995   
192    10B24E56-913B-4F1E-B5EC-0BCC1FC90859   
195    06B9F9C1-61DF-4469-BEB9-00CF3B5B3CD1   
198    19DC3525-FF1E-47ED-8CE5-0847EB4F30E3   
199    1D22A645-24C3-44C9-AA51-0CBEF3824DC8   
201    1F0B3520-6A74-484D-B78E-013E594C6CD2   
202    1F829180-B717-4BE4-9E09-0335378277E2   
203    1BC41F43-C34E-4DB4-B0E2-A619701A649C   
204    1F97D59E-5F53-4FB1-969C-04B180029CD2   
206    1F9EC60B-B5E6-48CF-9AFC-0582ABA04A84   
209    218C38F1-2CC2-4D8B-85C1-09113D1B8BD2   
216    32FAD83E-7812-4650-A4E2-039D0FB2A45E   
217    3312DB97-2F94-49AF-BE18-00B539D752F2   
219    36AF3D62-84B8-4F3B-A730-03F12A0D8E71   
221    37CDB5DA-DFAD-4919-937F-017FC5BF43EE   
225    3A3CB5B5-21AD-449A-97EC-048BCD66486E   

The list above shows that there are some very common erroneous entries in the organisation name field in Gateway to Research. We will assign all entries in the name column of the df DataFrame object as other = True and company = False, with the exception of 'unknown' and 'unlisted'.

In [14]:
org_counts = org_name_counter(df)
org_counts[(org_counts >= 5)]

unlisted                                 99
research                                 31
economics                                27
psychology                               25
history                                  24
chemistry                                21
physics                                  20
computer science                         19
english                                  17
geography                                17
law                                      17
philosophy                               17
education                                16
music                                    15
mathematics                              15
sociology                                14
grants administration                    13
politics                                 13
business school                          13
biological sciences                      13
research office                          12
unknown                                  12
finance                         

In [15]:
other_list = list(org_counts[org_counts >= 5].index)
other_list.remove('unlisted')
other_list.remove('unknown')

In [16]:
df.loc[(df.company == False) & (df.other == False), 'other'] = df[(df.company == False) &
                                                     (df.other == False)].name.isin(other_list)

In [17]:
print_stats(df)

Other:		4654
Companies:	11357
Unclassified:	9623
Clashes:	69


Let's just repeat this for those names appearing three time or more and less than 5 times

In [20]:
# Counts of each string occurence in name column
org_counts = org_name_counter(df)
org_counts[(org_counts >= 3) & (org_counts < 5)]

finance office                             4
materials                                  4
marketing                                  4
sch of modern languages                    4
research support services                  4
school of health sciences                  4
scottish association for marine science    4
administration                             4
anthropology                               4
psychiatry                                 4
sch of chemistry                           4
civil engineering                          4
institute of education                     4
school of chemistry                        4
school of physics and astronomy            4
biomedical sciences                        4
school of biological sciences              4
school of humanities                       4
biology                                    4
sch of social sciences                     4
architecture                               4
school of social sciences                  4
theology a

This time, there is one entry to exclude from the classification:
- BAE Systems (features 3 times)

In [21]:
other_list = list(org_counts[org_counts >= 3].index)
other_list.remove('unlisted')
other_list.remove('unknown')
other_list.remove('bae systems')

In [22]:
df.loc[(df.company == False) & (df.other == False), 'other'] = df[(df.company == False) &
                                                     (df.other == False)].name.isin(other_list)

In [23]:
print_stats(df)

Other:		4785
Companies:	11357
Unclassified:	9492
Clashes:	69


And once more, for org names appearing twice

In [24]:
# Counts of each string occurence in name column
org_counts = org_name_counter(df)
org_counts[(org_counts >= 2) & (org_counts < 3)]

department of physics                                2
art and design                                       2
denso corporation                                    2
materials engineering research laboratory limited    2
spend360 international limited                       2
london & scandinavian metallurgical co limited       2
medpharm ltd                                         2
mrc cognition and brain sciences unit                2
xceleron ltd                                         2
mrc toxicology unit                                  2
babraham institute                                   2
hispanic studies                                     2
university of oxford                                 2
stork technologies                                   2
criminology                                          2
management science                                   2
pipehawk plc                                         2
broad institute                                      2
agricultur

In [25]:
print('Num entries where name is >= 2 and < 3:\t{}'.format(
        org_counts[(org_counts >= 2) & (org_counts < 3)].sum()))

Num entries where name is >= 2 and < 3:	824


### Classifying with Companies House Data

At this point, manually assigning identifiers will become too problematic; there are a lot of strings to look at, and the differentiation between what is a company and what isn't is becoming less clear. An alternative method of classification for these remaining organisations is to use the Companies House API to compare organisation name strings against company names on their database. This will by no means classify all of the remainder of the organisations, but it may reduce the number remaining slightly. Specifically, we will try to classify organisations that have a specific UK postcode entry to be matched to Companies House data.

In [234]:
# first get a random sample that contain UK postcodes
# This is so we can easily test our appraoch on a smaller sample
df_rand = unclassified(df).sample(1000)
df_rand = df_rand[df_rand.postcodes.notnull()]
num_pc_entries = len(df_rand)
print("This run has {} random entries containing postcode information.".format(num_pc_entries))

This run has 619 random entries containing postcode information.


We use a simple process to define a match. First, the name of the organisation is taken from the dataframe, this is then passed as a search query to the Companies House API. If this search returns a match, then the postcode of that organisation is compared to the one returned by Companies House.

The caveat to this approach is that some organisations will have different postcodes registered with Companies House than they do on GtR.

In [235]:
df_rand['ch_json'] = df_rand.name.apply(search_org)
df_rand['ch_postcodes'] = df_rand.ch_json.apply(ch_postcodes)

In [245]:
df_rand.head()

id  \
12029  D91F10D1-0DFD-4A2E-9C0B-62C0B91EA840   
7367   703E38C8-1C6E-4EE3-9AFD-6CCA38A6F956   
204    1F97D59E-5F53-4FB1-969C-04B180029CD2   
7434   BCC12FEE-7A90-4AFE-8594-713906D8289E   
6363   A24EE84A-BAE3-4612-8E2B-085DD0056BC6   

                                           name  \
12029                           renew northwest   
7367            dr williams's trust and library   
204                         odgers conservation   
7434                      regen sw (south west)   
6363   british assoc for adoption and fostering   

                                               addresses company  other  \
12029  97183A3A-3B79-4952-8341-90263571A5B0 L1 4DQ Un...   False  False   
7367   C247F5BC-1489-4401-B53F-8A6D82009DC6 WC1H 0AR ...   False  False   
204    D38FD631-7F47-432E-9DF4-806FD0E8BE45 TA11 6BN ...   False  False   
7434   C8CFD016-12F7-4648-9277-26ED41B27A54 EX4 4RN M...   False  False   
6363   C57BD228-9433-49D5-842D-222DB35F0281 EC1N 8TS ...   False  False   

      postcodes                                            ch_json  \
12029    L1 4DQ  {'kind': 'search#companies', 'total_results': ...   
7367   WC1H 0AR  {'kind': 'search#companies', 'total_results': ...   
204    TA11 6BN  {'kind': 'search#companies', 'total_results': ...   
7434    EX4 4RN  {'kind': 'search#companies', 'total_results': ...   
6363   EC1N 8TS  {'kind': 'search#companies', 'total_results': ...   

      ch_postcodes  
12029    [OL9 7NZ]  
7367            []  
204      [BS1 5HH]  
7434            []  
6363      [B3 2BJ]

In [246]:
for index, row in df_rand.iterrows():
    df_rand.loc[index,'is_ch_company'] = row.postcodes in row.ch_postcodes

In [256]:
print('Out of {} organisations, {} were matched as companies.'.format(
        len(df_rand),
        len(df_rand[df_rand.is_ch_company == True])))

Out of 619 organisations, 134 were matched as companies.


We can now apply the same approach to the full dataframe

In [ ]:
df['ch_json'] = df.name.apply(search_org)
df['ch_postcodes'] = df.ch_json.apply(ch_postcodes)

In [9]:
for index, row in df.iterrows():
    df.loc[index,'is_ch_company'] = row.postcodes in row.ch_postcodes

In [10]:
print('Out of {} organisations, {} were matched as companies.'.format(
        len(df[(df.company == False) & (df.other == False)]),
        len(df[(df.is_ch_company == True) & (df.company == False) & (df.other == False)])))

Out of 9492 organisations, 1495 were matched as companies.


In [30]:
# Update company column with new matches
for index, row in df.iterrows():
     if row.company:
        df.loc[index, 'company'] = row.company
     elif not row.other:
        df.loc[index, 'company'] = row.is_ch_company
     else:
        df.loc[index, 'company'] = False
        
df.drop('is_ch_company', axis=1, inplace=True)
print_stats(df)

Other:		4785
Companies:	12852
Unclassified:	7997
Clashes:	69


Our final approach to organisational classification is to use the Open Corporates database. The reason for doing so is that the database contains nformation regarding dissolved companies, whereas Companies House do not currently release this via their API (though they may have plans to do so).

In [8]:
oc_key = os.environ["OPEN_CORPS_KEY"]

In [76]:
# Update company column with new matches
l = []
for index, row in df.iterrows():
    if not row.company and not row.other:
        r = search_open_corporates(row.name,
                                   country_code='gb',
                                   api_token=oc_key)
        l.append(r)
    else:
        l.append(False)

# New column from results list        
df['oc_results'] = pd.Series(l)

In [211]:
# Column of postcodes to match on
df['oc_postcodes'] = df.oc_results.apply(oc_postcodes)

In [212]:
for index, row in df.iterrows():
    df.loc[index,'is_oc_company'] = row.postcodes in row.oc_postcodes

In [252]:
df[df.is_oc_company == True]

Empty DataFrame
Columns: [id, name, addresses, company, other, postcodes, ch_json, ch_postcodes, oc_results, oc_postcodes, is_oc_company]
Index: []

The Open Corporates search didn't classify any of the remainder of the organisations. We now take a final step of checking the small number of clashes in the dataframe, where an organisation is classed as both a company and an organisation.

In [23]:
df[(df.company == True) & (df.other == True)]

id  \
557    E0242D92-6AA5-489E-93BF-647C834422C9   
940    59B462B7-8D1A-43F0-9E11-C8210A295255   
945    5154184E-5F5F-4E78-BDA7-C37CC089FC21   
1510   0D2FC2CE-DE17-4F8B-ACE2-32EB737D6387   
2100   85C0137D-EFE0-4C4F-A4C6-8B31C4F586A5   
2214   2B82458D-330C-4EB8-A6A5-EC8905115CEB   
2781   B020994B-44EA-4B36-8CF5-50ECC1785DD4   
2897   0D9CAEAF-C667-4EED-8AE3-C63A0D71E244   
3120   E6AE9487-83B8-4FEA-9E6B-94E1757F1287   
3341   BB919D63-84AF-416E-BE6C-85893EDD3BF4   
3760   DBC20C20-F980-4636-9450-C60D59AEEC2D   
4036   104C9D71-D2B9-4DD3-9F24-229634DCBAC0   
4082   724BC3C6-173C-466D-9324-22D202AE7A91   
4353   AFFFCE25-BC2E-4316-9B75-39C1AF79ADB5   
4658   F668A141-4E15-4227-99BC-4B9962C02EE0   
4753   1F8EF3BB-A8C6-4F84-AC2B-33F20784BBAC   
5692   E7EC83B5-171C-42FA-9B61-38AE77A6CA89   
6126   F9E880CD-0F3C-4AE0-AD6D-40203815D8E9   
6644   C90BE9A5-3C3E-42FC-A8F6-189440C49558   
6909   7BF981AB-34B9-491C-BB08-5678FB76CE31   
6942   D4DB533D-8F86-4BF0-BFA2-AC26784A896F   
7251   D0A07C2C-049E-4787-B8FC-3FC380E86725   
7689   1D9EC23A-3258-441C-9285-844C10C9DDD0   
7985   891A12CB-2C1E-4EAB-907E-C11180C8113A   
8670   AA8F777A-FA73-4399-83C4-BD77FC846D64   
8674   BB12030F-D36D-4D6F-A603-F4FB34CAA321   
8890   7977D55E-6CC4-40A1-9DB9-AFB6F5ED2608   
8895   B508089E-7426-4C0B-97B1-C63BC57672A1   
9132   C68D9A71-1AC0-4C1E-8C02-BCB0CB445AC8   
9430   EF7EA3B6-0CA5-402D-A86A-D0BD2103DAA9   
...                                     ...   
11488  81564991-FAC4-4F5E-B2CA-44AD7192F9F8   
12009  916A97F6-4C1A-4E27-9726-614F661E6650   
12507  A930F12E-4B01-4864-B3A8-4E3A731BD722   
13185  910D0D4E-5D06-4D40-953F-7E71966ABCFE   
13501  77F3096B-8798-43D1-A6CF-BD4AB0FD9B7E   
13830  B48A0015-944C-44BF-B5D8-DC8B9D047F6B   
14432  7CCBB886-38D6-4884-8062-FF726F854DD2   
14538  8CAEA975-7A6A-4720-AD0B-ED5FBC0B19C5   
14655  087B9388-D675-433D-9F2E-73DA7C8E11D1   
16450  40CF6C02-359E-4DD3-AF5A-4E98E1FCA82A   
18920  3F8D4336-F81C-4E5C-AEB7-D0508C037905   
19118  5DCF1104-2BA1-411F-A04F-6CA69EE25E4C   
19208  9CAEDF1A-DA5F-4444-9624-9F68115989CA   
19416  EB648E66-0418-4D15-8088-805FFED9824F   
19960  ECECFB77-1371-4FFA-B2BC-939AD230F161   
20196  6315F388-0590-42D3-8E27-C1DE006B6000   
20768  B3F7ACFF-0061-49E8-A25C-0FEE21AD2695   
21181  E7EC7F4B-F0DA-4238-B2A4-2AFCA4E83ECD   
21213  DA56F4CC-1725-44D2-8DF3-2DD28FE1CB51   
21581  2220746F-CAF8-41B9-A3BB-5767020B26E8   
21947  97B642A5-21B5-4B10-9E58-2E9FB7017915   
22264  5D5BD5C3-BCAE-4F77-B60B-3C4C633B8302   
22848  90131B73-31F8-46A4-8396-4DDD5EDC19C2   
22976  BE531B9A-CC1F-4294-9CCC-873454473CFA   
23998  1BA799C3-B34F-413B-A62C-5FD3E0C9FEB5   
24094  D2A63A6C-A98A-425F-9D8E-BB7AE974D9B5   
24261  315B7FC9-5D3B-4670-87B9-7A31E9471C20   
24694  D3AB18F9-6746-45C2-A2AB-7AB16DC6F4F7   
24901  8C362313-83DF-432F-8700-FBB396CDC0A3   
25477  85AC9DEA-7D7E-47CF-8C74-CB60A23865BA   

                                                    name  \
557                                   sphere science ltd   
940                     gardline marine sciences limited   
945                            success at school limited   
1510                         linnaeus plant sciences inc   
2100                               field science limited   
2214                  the capita group plc / nhs choices   
2781                     elife sciences publications ltd   
2897                                   t-ray science inc   
3120                            science city bristol ltd   
3341                                science practice ltd   
3760              university of warwick science park ltd   
4036                               sei oxford office ltd   
4082                            tg life sciences limited   
4353                                agt sciences limited   
4658                         ultrasonic sciences limited   
4753                   buffs social club & institute ltd   
5692         alterity business and academic services ltd   
6126                    eurotek office furnitu

In [39]:
other = [1375, 10225]
mask = df['other'].isin(other)
df.loc[(~mask) & (df.company == True) & (df.other == True), 'other'] = False

In [43]:
for x in other:
    df.set_value(x, 'company', False) 

In [44]:
print_stats(df)

Other:		4718
Companies:	12850
Unclassified:	7997
Clashes:	0


In [45]:
# Save to pickle for easy loading
df.to_pickle('df_organisation_type.p')

In [47]:
# Read in config file with DB params
with open('../scripts/config.json') as f:
    conf = json.load(f)
    
engine = create_engine('postgresql://{}:{}@{}/{}'.format(conf.get('user'),
                                                         conf.get('passw'),
                                                         conf.get('host'),
                                                         conf.get('database')))

conn = engine.connect()

df[['id', 'company', 'other']].to_sql('orgs_type',
                                      conn,
                                      schema='gtr',
                                      if_exists='replace',
                                      index=False)